addons/survey/models/survey_survey.py

```diff
@@ -397,6 +408,13 @@ def _onchange_session_speed_rating(self):
             survey.question_ids._update_time_limit_from_survey(
                 is_time_limited=survey.session_speed_rating, time_limit=survey.session_speed_rating_time_limit)
 
+    @api.onchange('restrict_user_ids')
+    def _onchange_restrict_user_ids(self):
+        for survey in self.filtered(lambda s: s.restrict_user_ids and bool(
+            s.user_id - s.restrict_user_ids) and not s.user_id.has_group(
+                'survey.group_survey_manager')):
+            survey.restrict_user_ids += survey.user_id
```

Comment:
We should only check once per user
```suggestion
        surveys_to_check = self.filtered(lambda s: s.restrict_user_ids and bool(s.user_id - s.restrict_user_ids))
        users_are_managers = {
            user: user.has_group('survey.group_survey_manager') 
            for user in surveys_to_check.user_id
        }
        for survey in surveys_to_check.filtered(lambda s: not users_are_managers[s.user_id]):
            survey.restrict_user_ids += survey.user_id
```